contracts/interchain-token-service/src/contract.rs

```diff
@@ -329,380 +183,146 @@ impl AxelarExecutableInterface for InterchainTokenService {
-            assert_eq!(message, msg);
+fn get_execute_params(
+    env: &Env,
+    source_chain: String,
+    payload: &Bytes,
+) -> Result<(String, Message), ContractError> {
+    let message_type =
+        get_message_type(&payload.to_alloc_vec()).map_err(|_| ContractError::InvalidPayload)?;
+
+    match message_type {
+        EncodedMessageType::ReceiveFromHub => {
+            ensure!(
+                source_chain == InterchainTokenService::its_hub_chain_name(env),
+                ContractError::UntrustedChain
+            );
+
+            let decoded_message =
+                HubMessage::abi_decode(env, payload).map_err(|_| ContractError::InvalidPayload)?;
+
+            let HubMessage::ReceiveFromHub {
+                source_chain: original_source_chain,
+                message: inner_message,
+            } = decoded_message
+            else {
+                return Err(ContractError::InvalidMessageType);
+            };
+
+            let trusted_address =
+                InterchainTokenService::trusted_address(env, original_source_chain.clone());
+            let routing_identifier = InterchainTokenService::its_hub_routing_identifier(env);
+
+            ensure!(
+                trusted_address.is_some_and(|addr| addr == routing_identifier),
+                ContractError::UntrustedChain
+            );
+
+            Ok((original_source_chain, inner_message))
         }
+        _ => Err(ContractError::InvalidMessageType),
     }
+}
 
-    #[test]
-    fn get_execute_params_fails_hub_message_sent_from_external_chain() {
-        let (env, client) = setup_env();
-        register_chains(&env, &client);
-
-        let msg = HubMessage::ReceiveFromHub {
-            source_chain: String::from_str(&env, HUB_CHAIN),
-            message: Message::DeployInterchainToken(DeployInterchainToken {
-                token_id: BytesN::from_array(&env, &[1u8; 32]),
-                name: String::from_str(&env, &"Test Token"),
-                symbol: String::from_str(&env, &"TST"),
-                decimals: 18,
-                minter: Some(bytes_from_hex(&env, "1234")),
-            }),
-        }
-        .abi_encode(&env);
-
-        let result = env.as_contract(&client.address, || {
-            InterchainTokenService::get_execute_params(
-                &env,
-                String::from_str(&env, "somechain"),
-                msg.unwrap(),
-            )
-        });
-        assert!(matches!(result, Err(ContractError::UntrustedChain)));
-    }
+fn get_call_params(
+    env: &Env,
+    destination_chain: String,
+    message: Message,
+) -> Result<Bytes, ContractError> {
+    // Note: ITS Hub chain as the actual destination chain for the messsage isn't supported
+    ensure!(
+        destination_chain != InterchainTokenService::its_hub_chain_name(env),
+        ContractError::UntrustedChain
+    );
+
+    match InterchainTokenService::trusted_address(env, destination_chain.clone()) {
```

Comment:
```suggestion
    let Some(destination_address) = InterchainTokenService::trusted_address(env, destination_chain.clone()) else { return Err(...) };
```
use let else instead of match